src/esp/core/Configuration.h

```diff
@@ -9,184 +9,649 @@
+   */
+  void setSubconfigPtr(const std::string& name,
+                       std::shared_ptr<Configuration>& configPtr) {
+    configMap_[name] = std::move(configPtr);
+  }  // setSubconfigPtr
 
-// Below uses std::variant; not yet available in clang c++17
-/*
-#include <std/variant>
+  int getNumSubconfigs(const std::string& name) const {
+    if (configMap_.count(name) > 0) {
+      return configMap_.at(name)->getNumEntries();
+    }
+    ESP_WARNING() << "No Subconfig found named :" << name;
+    return 0;
+  }
 
-class Configuration {
- public:
-  Configuration() : values_() {}
+  /**
+   * @brief Return number of entries in this configuration. This only counts
+   * subconfiguration entries as a single entry.
+   */
+  int getNumEntries() const { return configMap_.size() + valueMap_.size(); }
 
-  template <typename T>
-  bool set(const std::string& key, const T& value) {
-    bool didOverride = hasValue(key);
-    values_[key] = value;
-    return didOverride;
+  bool hasValues() const { return getNumEntries() > 0; }
+
+  /**
+   * @brief Returns whether this @ref Configuration has the passed @p key as a
+   * non-configuration value. Does not check subconfigurations.
+   */
+  bool hasValue(const std::string& key) const {
+    return valueMap_.count(key) > 0;
   }
 
-  template <typename T>
-  T get(const std::string& key) const {
-    return boost::get<T>(values_.at(key));
+  /**
+   * @brief Checks if passed @p key is contained in this configuration.
+   * Returns a list of nested subconfiguration keys, in order, to the
+   * configuration where the key was found, ending in the requested @p key. If
+   * list is empty, @p key was not found.
+   * @param key The key to look for
+   * @return A breadcrumb list to where the value referenced by @p key
+   * resides. An empty list means the value was not found.
+   */
+  std::vector<std::string> findValue(const std::string& key) const {
+    std::vector<std::string> breadcrumbs{};
+    findValueInternal(key, 0, breadcrumbs);
+    return breadcrumbs;
   }
 
-  bool hasValue(const std::string& key) {
-    return values_.count(key) > 0;
+  /**
+   * @brief Builds and returns @ref Corrade::Utility::ConfigurationGroup
+   * holding the values in this esp::core::Configuration.
+   *
+   * @return a reference to a configuration group for this configuration
+   * object.
+   */
+  Corrade::Utility::ConfigurationGroup getConfigGroup() const {
+    Corrade::Utility::ConfigurationGroup cfg{};
+    putAllValuesInConfigGroup(cfg);
+    return cfg;
+  }
+
+  /**
+   * @brief This method will build a vector of all the config values this
+   * configuration holds and the types of these values.
+   */
+  std::unordered_map<std::string, ConfigStoredType> getValueTypes() const {
+    std::unordered_map<std::string, ConfigStoredType> res{};
+    res.reserve(valueMap_.size());
+    for (const auto& elem : valueMap_) {
+      res[elem.first] = elem.second.getType();
+    }
+    return res;
+  }
```

Comment:
This function was added for informational purposes, primarily to provide a listing of all the available keys.  Since, with the user-defined attributes configuration, the user may not remember all the keys that are currently available in the Configuration.  Originally, this was a vector of strings, but the types were added to provide extra information to the user.